tools/testing/gatekeeper/jobs.py

```
@@ -0,0 +1,189 @@
+        """Get jobs from a workflow id"""
+        total_count = -1
+        jobs = []
+        page = 1
+        while True:
+            response = requests.get(
+                f"{_GH_RUNS_URL}/{run_id}/jobs?per_page=100&page={page}",
+                headers=_GH_HEADERS,
+                timeout=60
+            )
+            response.raise_for_status()
+            output = response.json()
+            jobs.extend(output["jobs"])
+            total_count = max(total_count, output["total_count"])
+            if len(jobs) >= total_count:
+                break
+            page += 1
+        return jobs
+
+    def check_workflow_runs_status(self):
+        """
+        Checks if all required jobs passed
+
+        :returns: 0 - all passing; 1 - any failure; 127 some jobs running
+        """
+        # TODO: Check if we need pagination here as well
+        latest_commit_sha = os.getenv("COMMIT_HASH")
+        response = requests.get(
+            _GH_RUNS_URL,
+            params={"head_sha": latest_commit_sha},
+            headers=_GH_HEADERS,
+            timeout=60
+        )
+        response.raise_for_status()
+        workflow_runs = response.json()["workflow_runs"]
+
+        for run in workflow_runs:
+            jobs = self.get_jobs_for_workflow_run(run["id"])
+            for job in jobs:
+                self.record(run["name"], job)
+        print(self)
+        return self.status()
+
+    def run(self):
+        """
+        Keep checking the PR until all required jobs finish
+
+        :returns: 0 on success; 1 on failure
+        """
+        while True:
+            ret = self.check_workflow_runs_status()
+            if ret == RUNNING:
+                running_jobs = len([job
+                                    for job, status in self.results
+                                    if status == RUNNING])
+                print(f"{running_jobs} jobs are still running...")
+                time.sleep(60)
```

It would be good to parameterize this value, 60s seems too small (build and test jobs take several minutes to finish). I.e. we might need to tune that value to ensure we don't hit the API rate limits:

 * https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-github_token-in-github-actions
 * https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits

---

Wow, you found the limits, good. We can change it as well as we can make it dynamic (let's say we expect the failures first after 15 or 30m, then we can start checking per 1m and after 2h increase it to hour (the assumption would be we know tests don't finish earlier than in 30m, then we want to be checking frequently to fail fast but when things take too long, we don't need that frequency and can check once an hour). Or anything else we can come-up with.

---

Anyway, I bumped it to 3m for now, let me know if you want some advanced heuristic

---

Initially 3m seems fine. let's see how it behave in practice ;)